tests/core/featurizers/test_precomputation.py

```diff
@@ -0,0 +1,565 @@
+
+
+def test_precomputation_lookup_features_if_lookup_table_is_broken():
+
+    broken_table = CoreFeaturizationPrecomputations()
+    broken_table._table = {
+        CoreFeaturizationPrecomputations._build_key({TEXT: "A"}): Message(
+            data={}, features=[_dummy_features(2, TEXT)]
+        ),
+        CoreFeaturizationPrecomputations._build_key({INTENT: "B"}): Message(
+            data={}, features=[_dummy_features(1, TEXT)]
+        ),
+    }
+    with pytest.raises(
+        RuntimeError, match=f"Feature for attribute {TEXT} has already been"
+    ):
+        broken_table.lookup_features({TEXT: "A", INTENT: "B"})
+
+    # as a reminder that we don't fix every strange edgecase:
+    not_broken_but_strange_table = CoreFeaturizationPrecomputations()
+    not_broken_but_strange_table._table = {
+        CoreFeaturizationPrecomputations._build_key({TEXT: "A"}): Message(data=dict()),
+        CoreFeaturizationPrecomputations._build_key({INTENT: "B"}): Message(
+            data=dict(), features=[_dummy_features(1, TEXT)]
+        ),
+    }
+    features = not_broken_but_strange_table.lookup_features({TEXT: "A", INTENT: "B"})
+    assert TEXT in features and len(features[TEXT]) == 1
+
+
+def test_precomputation_lookup_message():
+
+    OTHER = "other"
+    messages = [
+        Message(data={TEXT: "A"}, features=[_dummy_features(1, TEXT)]),
+        Message(data={TEXT: "B"}),
+        Message(data={TEXT: "B", OTHER: "C"}, features=[_dummy_features(2, OTHER)]),
+        Message(data={INTENT: "B"}),
+        Message(data={ACTION_TEXT: "B"}),
+        Message(data={ACTION_NAME: "B"}),
+    ]
+
+    table = CoreFeaturizationPrecomputations(handle_collisions=True)
+    table.add_all(messages)
+
+    message = table.lookup_message(user_text="A")
+    assert message
+    message = table.lookup_message(user_text="B")
+    assert OTHER in message.data.keys()
+    with pytest.raises(ValueError, match=f"Expected a message with key \\('{TEXT}"):
+        table.lookup_message(user_text="not included")
+
+
+def test_precomputation_derive_messages_from_events_and_add():
+    events = [
+        UserUttered(intent={INTENT_NAME_KEY: "greet"}),
+        ActionExecuted(action_name="utter_greet"),
+        ActionExecuted(action_name="utter_greet"),
+    ]
+    lookup_table = CoreFeaturizationPrecomputations(handle_collisions=False)
+    lookup_table.derive_messages_from_events_and_add(events)
+    # should create: one intent-only user substate and one prev_action substate
+    assert len(lookup_table) == 2
+
+    events = [
+        UserUttered(text="text", intent={INTENT_NAME_KEY: "greet"}),
+        ActionExecuted(action_name="utter_greet"),
+    ]
+    lookup_table = CoreFeaturizationPrecomputations(handle_collisions=False)
+    lookup_table.derive_messages_from_events_and_add(events)
+    # should create: one intent-only user substate, one text-only user substate one
+    # prev_action substate
+    assert len(lookup_table) == 3
```

Comment:
how about using `pytest.mark.parametrize` for this?

Reply:
good idea, yep :)